### PR TITLE
[DOC] Change snippets for cigar to use the cigar alphabet instead of …

### LINKED ISSUE #1216: Cigar documentation uses snippets from qualfied alphabet
## Platform

  - SeqAn version: master

## Description

The documentation of the cigar alphabet shows examples for the qualified alphabet in many member function documentations.

### PR DESCRIPTION
**Pull request by joshuak94**
…qualified. Also remove the subtype snippets as it is not applicable.

Fixes #1216

### CODE REVIEW

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 1]
```diff
@@ -72,14 +72,23 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
     // Inherit constructors from base
     using base_t::base_t;
 
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(component_type const alph)
+    /*!\brief Construction via a value of one of the components.
+     * \tparam component_type Must be one uniquely contained in the type
+                              list of the composite.
+     * \param  alph           The value of a component that should be assigned.
```
Comment (smehringer) [inline include/seqan3/alphabet/cigar/cigar.hpp]: why can't you use the \copydoc anymore? You can use copydoc AND give more details :)

File: test/snippet/alphabet/cigar/cigar_op.cpp [hunk 2]
```diff
@@ -5,9 +5,9 @@ using seqan3::operator""_cigar_op;
 
 int main()
 {
-// Initialze an seqan3::cigar_op:
-seqan3::cigar_op match{'M'_cigar_op};
+    // Initialze an seqan3::cigar_op:
```
Comment (smehringer) [inline test/snippet/alphabet/cigar/cigar_op.cpp]:
```suggestion
    // Initialise a seqan3::cigar_op:
```

File: test/snippet/alphabet/cigar/cigar_value_construction.cpp [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+{
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter1{0};
+    // creates 0M, as the cigar_op field is not provided.
+    seqan3::cigar letter2{'M'_cigar_op};
+    // creates 0M, as the integer field is not provided.
+
+    if (letter1 == letter2)
+    seqan3::debug_stream << "yeah\n"; // yeah
```
Comment (smehringer) [inline test/snippet/alphabet/cigar/cigar_value_construction.cpp]:
```suggestion
        seqan3::debug_stream << "yeah\n"; // yeah
```

File: include/seqan3/alphabet/cigar/cigar.hpp (hunk 1 above)
Comment (joshuak94) [reply include/seqan3/alphabet/cigar/cigar.hpp]: Hmmm I guess I could! That would also copy the qualified snippet over, though. Then there would be the qualified snippet example *and* the cigar example, and I think the reason for the issue was that the qualified one shouldn't be in the cigar documentation?

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 4]
```diff
@@ -72,14 +72,23 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
     // Inherit constructors from base
     using base_t::base_t;
 
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(component_type const alph)
+    /*!\brief Construction via a value of one of the components.
+     * \tparam component_type Must be one uniquely contained in the type
+                              list of the composite.
```
Comment (smehringer) [inline include/seqan3/alphabet/cigar/cigar.hpp]: This line should still fit in the line above. Same in the other documentation.

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 5]
```diff
@@ -72,14 +72,21 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
     // Inherit constructors from base
     using base_t::base_t;
 
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(component_type const alph)
+    /*!\brief Construction via a value of one of the components.
+     * \tparam component_type One of the component types. Must be uniquely contained in the type list of the composite.
+     * \param  alph           The value of a component that should be assigned.
```
Comment (rrahn) [inline include/seqan3/alphabet/cigar/cigar.hpp]:
```suggestion
     * \param[in]  alph           The value of a component that should be assigned.
```

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 6]
```diff
@@ -72,14 +72,21 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
...
+     * \include test/snippet/alphabet/cigar/cigar_value_construction.cpp
+     */
     SEQAN3_DOXYGEN_ONLY(( constexpr cigar(component_type const alph) noexcept {} ))
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(indirect_component_type const alph)
-    SEQAN3_DOXYGEN_ONLY(( constexpr cigar(indirect_component_type const alph) noexcept {} ))
-    //!\copydoc alphabet_tuple_base::operator=(component_type const alph)
+
+    /*!\brief Assignment via a value of one of the components.
+     * \tparam component_type One of the component types. Must be uniquely contained in the type list of the composite.
+     * \param  alph           The value of a component that should be assigned.
```
Comment (rrahn) [inline include/seqan3/alphabet/cigar/cigar.hpp]:
```suggestion
     * \param[in] alph           The value of a component that should be assigned.
```

**Comment by joshuak94** [comment]
@rrahn I just tried it, it doesn't work.

**Comment by rrahn** [comment]
> @rrahn I just tried it, it doesn't work.

Ok, might make sense, since it can be hard if the type is the same. But then there should be some documentation about `get<0>(letter1)`, respectively `get<int32_t>(letter1)`. I think this is crucial for people dealing with the cigar alphabet.

**Comment by joshuak94** [comment]
Where should that documentation go? I guess I can do this SEQAN3_DOXYGEN_ONLY thing right? Like this:

```cpp
/*!\copydoc alphabet_tuple_base::get(alphabet_tuple_base & l)
 * \include test/snippet/alphabet/cigar/get_index.cpp
 */
SEQAN3_DOXYGEN_ONLY(( template <size_t index> constexpr auto get(cigar & l) noexcept {} ))

/*!\copydoc alphabet_tuple_base::get(alphabet_tuple_base & l)
 * \include test/snippet/alphabet/cigar/get_type.cpp
 */
SEQAN3_DOXYGEN_ONLY(( template <typename type> constexpr auto get(cigar & l) noexcept {} ))
```

But, how does copydoc know which get is which?

**Comment by smehringer** [comment]
@rrahn

> `size_t count = letter1`

I think would be quite hacky because you would need to implement user defined conversion operators, based on each template argument by pack expansion. And they are hacky already.
Also this would enable implicit casts correct? Which I think is error prone.

@joshuak94

> But, how does copydoc know which get is which?

Try it out :) Otherwise you have to document one of them manually.

**Comment by joshuak94** [comment]
@smehringer I've submitted a first attempt at the snippets, can you re-review?

**Comment by rrahn** [comment]
> I think would be quite hacky because you would need to implement user defined conversion operators, based on each template argument by pack expansion. And they are hacky already.
Also this would enable implicit casts correct? Which I think is error prone.

I know. I just wasn't sure if there is already something like this, because it is a nice user interface. Better than the get interface which forces me to remember the order. Also this might be natural I think adding getter functions in the special case of cigar would have an improved usability for the user. At least I find `letter1.operation_size()` and `letter1.operation_type()` easier to read then `get<0>(letter1)` and `get<1>(letter1)`. Although, `get<int32_t>(letter1)` and `std::get<seqan3::cigar_op>(letter1)` would be fine, it still comes with the burden of using `using std::get` before.

**Comment by smehringer** [comment]
> At least I find `letter1.operation_size()` and `letter1.operation_type()` easier to read

we can discuss to additionally provide those, but I think get is fair given that this is an `alphabet_tuple`?
Shall I add a discussion card?

File: test/snippet/alphabet/cigar/cigar_get_index.cpp [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+#include <seqan3/core/debug_stream.hpp>
+
+int main()
+{
+    using seqan3::get;
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter0{10, 'M'_cigar_op};
+
+    uint32_t               size{get<0>(letter0)}; // Note this is equivalent to get<uint32_t>(letter0)
```
Comment (smehringer) [inline test/snippet/alphabet/cigar/cigar_get_index.cpp]:
```suggestion
    uint32_t size{get<0>(letter0)};               // Note this is equivalent to get<uint32_t>(letter0)
```

File: test/snippet/alphabet/cigar/cigar_get_type.cpp [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+#include <seqan3/core/debug_stream.hpp>
+
+int main()
+{
+    using seqan3::get;
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter0{10, 'M'_cigar_op};
+
+    uint32_t               size{get<uint32_t>(letter0)};         // Note this is equivalent to get<0>(letter0)
```
Comment (smehringer) [inline test/snippet/alphabet/cigar/cigar_get_type.cpp]:
```suggestion
    uint32_t size{get<uint32_t>(letter0)};                       // Note this is equivalent to get<0>(letter0)
```

**Comment by rrahn** [comment]
> Shall I add a discussion card?

Yes, please

File: test/snippet/alphabet/cigar/cigar_get_type.cpp [hunk 9]
```diff
@@ -0,0 +1,18 @@
+#include <seqan3/alphabet/cigar/cigar.hpp>
+#include <seqan3/alphabet/cigar/cigar_op.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+int main()
+{
+    using seqan3::get;
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter0{10, 'M'_cigar_op};
```
Comment (rrahn) [inline test/snippet/alphabet/cigar/cigar_get_type.cpp]:
```suggestion
    seqan3::cigar letter{10, 'M'_cigar_op};
```
There is no other letter.

File: test/snippet/alphabet/cigar/cigar_get_index.cpp [hunk 10]
```diff
@@ -0,0 +1,18 @@
+#include <seqan3/alphabet/cigar/cigar.hpp>
+#include <seqan3/alphabet/cigar/cigar_op.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+int main()
+{
+    using seqan3::get;
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter0{10, 'M'_cigar_op};
```
Comment (rrahn) [inline test/snippet/alphabet/cigar/cigar_get_index.cpp]:
```suggestion
    seqan3::cigar letter{10, 'M'_cigar_op};
```

there is no other letter.

File: test/snippet/alphabet/cigar/cigar_value_assignment.cpp [hunk 11]
```diff
@@ -0,0 +1,16 @@
+#include <seqan3/alphabet/cigar/cigar.hpp>
+#include <seqan3/alphabet/cigar/cigar_op.hpp>
+#include <seqan3/core/debug_stream.hpp>
+
+int main()
+{
+    using seqan3::operator""_cigar_op;
+
+    seqan3::cigar letter1{10, 'I'_cigar_op};
```
Comment (rrahn) [inline test/snippet/alphabet/cigar/cigar_value_assignment.cpp]:
```suggestion
    seqan3::cigar letter{10, 'I'_cigar_op};
```
there is no other letter

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 12]
```diff
@@ -133,6 +140,24 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
...
+     */
+    /*!\copydoc alphabet_tuple_base::get(alphabet_tuple_base & l)
+     *
+     * \include test/snippet/alphabet/cigar/cigar_get_index.cpp
+     */
+    SEQAN3_DOXYGEN_ONLY(( template <size_t index> constexpr auto get(cigar & l) noexcept {} ))
+
+    /*!\copybrief get
+     * \tparam type Return the element of specified type; only available if the type is unique in the set of components.
+     * \returns A proxy to the contained element that models the same alphabets concepts and supports assignment.
```
Comment (rrahn) [inline include/seqan3/alphabet/cigar/cigar.hpp]:
```suggestion
     * \returns A proxy to the contained element that models the same alphabet concept and supports assignment.
```

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 13]
```diff
@@ -72,14 +72,21 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
...
+     *
+     * \include test/snippet/alphabet/cigar/cigar_value_construction.cpp
+     */
     SEQAN3_DOXYGEN_ONLY(( constexpr cigar(component_type const alph) noexcept {} ))
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(indirect_component_type const alph)
-    SEQAN3_DOXYGEN_ONLY(( constexpr cigar(indirect_component_type const alph) noexcept {} ))
-    //!\copydoc alphabet_tuple_base::operator=(component_type const alph)
+
+    /*!\brief Assignment via a value of one of the components.
+     * \tparam component_type One of the component types. Must be uniquely contained in the type list of the composite.
```
Comment (rrahn) [inline include/seqan3/alphabet/cigar/cigar.hpp]:
```suggestion
     * \tparam component_type One of the component types; must be uniquely contained in the type list of the composite.
```

File: include/seqan3/alphabet/cigar/cigar.hpp [hunk 14]
```diff
@@ -72,14 +72,21 @@ class cigar : public alphabet_tuple_base<cigar, uint32_t, cigar_op>
     // Inherit constructors from base
     using base_t::base_t;
 
-    //!\copydoc alphabet_tuple_base::alphabet_tuple_base(component_type const alph)
+    /*!\brief Construction via a value of one of the components.
+     * \tparam component_type One of the component types. Must be uniquely contained in the type list of the composite.
```
Comment (rrahn) [inline include/seqan3/alphabet/cigar/cigar.hpp]:
```suggestion
     * \tparam component_type One of the component types; must be uniquely contained in the type list of the composite.
```

**Pull request merged**